src/HTSPDemuxer.cpp

```diff
@@ -630,21 +654,127 @@ void CHTSPDemuxer::ParseSubscriptionSpeed ( htsmsg_t *m )
 
 void CHTSPDemuxer::ParseSubscriptionStatus ( htsmsg_t *m )
 {
-  const char *status;
+  const char *status, *error;
   status = htsmsg_get_str(m, "status");
+  error  = htsmsg_get_str(m, "subscriptionError");
 
-  // not for preTuning subscriptions
-  if (m_subscription.weight != SUBSCRIPTION_WEIGHT_DEFAULT)
+  // not for preTuning subscriptions, tag them as unknown
+  if (m_subscription.weight < SUBSCRIPTION_WEIGHT_DEFAULT)
+  {
+    m_subscription.state = SUBSCRIPTION_UNKNOWN;
     return;
+  }
 
   // this field is absent when everything is fine
   if (status != NULL)
-  {
     tvhinfo("Bad subscription status: %s", status);
-    XBMC->QueueNotification(QUEUE_INFO, status);
+
+  /* 'subscriptionError' was added in htsp v20 */
+  /* Use 'status' for older tvheadend backends */
+  if (m_conn.GetProtocol() >= 20)
+  {
+    if (error != NULL)
+    {
+      if (!strcmp("badSignal", error))
+        m_subscription.state = SUBSCRIPTION_NOSIGNAL;
+      else if (!strcmp("scrambled", error))
+        m_subscription.state = SUBSCRIPTION_SCRAMBLED;
+      else if (!strcmp("userLimit", error))
+        m_subscription.state = SUBSCRIPTION_USERLIMIT;
+      else if (!strcmp("noFreeAdapter", error))
```

Comment:
Hmm. Another piece of code, same story, same design flaw, if you ask me.

Why do you not just return SM_CODE in the tvh API? This would be a much better approach, IMO. What you do here is convert the stuff you translated  in tvh from numerical values into strings back into numerical values (maybe even the same numerically as used in tvh - not checked).

Anyhow, it is like it is now as the actual flaw is in tvh. Maybe you take into account what I said (if you can follow my arguments at all) when you add stuff to the tvh API next time.